### PR TITLE
MULE-14303: Nested processor executes start before initialise when in…

### PR DESCRIPTION
**Pull request by alegmarra**
…side a router route

### CODE REVIEW

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ComponentMessageProcessorObjectFactory.java [hunk 1]
```diff
@@ -63,7 +63,8 @@ public P doGetObject() throws Exception {
       componentModel.getNestedComponents().stream()
           .filter(component -> component instanceof NestedChainModel)
           .findFirst()
-          .ifPresent(chain -> parameters.put(chain.getName(), new ProcessorChainValueResolver(nestedProcessors)));
+          .ifPresent(chain -> parameters.put(chain.getName(),
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ComponentMessageProcessorObjectFactory.java]: superflous

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/construct/RouteComponentObjectFactory.java [hunk 2]
```diff
@@ -60,7 +60,8 @@ public Object doGetObject() throws Exception {
         model.getNestedComponents().stream()
             .filter(component -> component instanceof NestedChainModel)
             .findFirst()
-            .ifPresent(chain -> parameters.put(chain.getName(), new ProcessorChainValueResolver(nestedProcessors)));
+            .ifPresent(chain -> parameters.put(chain.getName(),
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/construct/RouteComponentObjectFactory.java]: superflous

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/ExtensionComponent.java [hunk 3]
```diff
@@ -154,6 +155,7 @@ protected ExtensionComponent(ExtensionModel extensionModel,
   @Override
   public final void initialise() throws InitialisationException {
     if (cursorProviderFactory == null) {
+      initialiseIfNeeded(streamingManager, muleContext);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/ExtensionComponent.java]: in which case isn't the streamingManager initialised? The runtime controls the lifecycle of this guy. This doesn't seem right

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/objectbuilder/DefaultObjectBuilder.java [hunk 4]
```diff
@@ -95,4 +104,10 @@ public void setName(String name) {
   public void setEncoding(String encoding) {
     this.encoding = encoding;
   }
+
+  @Override
+  public void initialise() throws InitialisationException {
+    initialiseIfNeeded(resolvers.values(), muleContext);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/objectbuilder/DefaultObjectBuilder.java]: Shouldn't this lead to the resolvers being initialised twice?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java [hunk 5]
```diff
@@ -71,16 +74,9 @@ public boolean isDynamic() {
...
-        registerObject(muleContext, new ObjectNameHelper(muleContext).getUniqueName(""), chain);
-        LifecycleUtils.initialiseIfNeeded(chain, muleContext);
-      } catch (Exception e) {
-        throw new MuleRuntimeException(createStaticMessage("Could not register nested operation message processor"), e);
-      }
-      initialised.set(true);
-    }
+  @Override
+  public void initialise() throws InitialisationException {
+    initialiseIfNeeded(chain, muleContext);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java]: it doesn't make sense to do this here if the chain was registered before the initialisation phase (unless the constructor is invoked during the initialisation phase). All of this is very shady either way. Review, fix and/or clarify

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverUtils.java [hunk 6]
```diff
@@ -139,36 +140,38 @@ public static Object resolveCursor(Object value) {
...
       } else {
         TypeSafeValueResolverWrapper typeSafeValueResolverWrapper =
             new TypeSafeValueResolverWrapper<>(new StaticValueResolver<>(expression), getType(type));
         typeSafeValueResolverWrapper.setTransformationService(muleContext.getTransformationService());
-        typeSafeValueResolverWrapper.initialise();
-        return typeSafeValueResolverWrapper;
+        resolver = typeSafeValueResolverWrapper;
       }
+
+      initialiseIfNeeded(resolver, muleContext);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverUtils.java]: again, pretty sure this resolver will be initialised again at some point. Additionally, I get the impression you're calling this full initialisation method when all you really want is to perform DI.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/VoidCallbackArgumentResolver.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import org.mule.runtime.module.extension.api.runtime.privileged.ExecutionContextAdapter;
+import org.mule.runtime.module.extension.internal.ExtensionProperties;
+
+/**
+ * {@link ArgumentResolver} which returns the {@link ExtensionProperties#COMPLETION_CALLBACK_CONTEXT_PARAM}
+ * context variable.
+ * <p>
+ * Notice that this resolver only works if the {@link ExecutionContext} is a {@link ExecutionContextAdapter}
+ *
+ * @since 4.0
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/VoidCallbackArgumentResolver.java]: 4.1

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ComponentMessageProcessorObjectFactory.java (hunk 1 above)
Comment (alegmarra) [reply modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ComponentMessageProcessorObjectFactory.java]: adding the `muleContext` as constructor parameter, that's the reason of the line break

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/construct/RouteComponentObjectFactory.java (hunk 2 above)
Comment (alegmarra) [reply modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/construct/RouteComponentObjectFactory.java]: added the `muleContext` as constructor parameter, that's the reason of the line break

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverUtils.java (hunk 6 above)
Comment (alegmarra) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverUtils.java]: some were being initialised and some did not. This is normalization

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java [hunk 8]
```diff
@@ -33,18 +33,21 @@
...
   private final MessageProcessorChain chain;
 
-
-  @Inject
-  private MuleContext muleContext;
-
-  public ProcessorChainValueResolver(MuleContext ctx, MessageProcessorChain chain) {
-    this.chain = chain;
+  public ProcessorChainValueResolver(MuleContext ctx, final MessageProcessorChain delegate) {
+    this.chain = (MessageProcessorChain) Enhancer.create(MessageProcessorChain.class,
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ProcessorChainValueResolver.java]: not... half moon

**Pull request merged**